Background Task Runner example (#1802)
umbraco/UmbracoDocs: Reference/Scheduling/index.md

---
versionFrom: 8.0.0
---

# Scheduling with BackgroundTaskRunner
In Umbraco 8+ it is possible to run reccuring code using the `BackgroundTaskRunner`.

> Below is a complete example showing how to register a Task with the runner with a component that will regularly empty out the recycle bin every five minutes.

Reviewer: ```suggestion
Below is a complete example showing how to register a Task Runner with a [component](../../Implementation/Composing/index.md) that will regularly empty out the recycle bin every five minutes.
```

```csharp
using Umbraco.Core;
using Umbraco.Core.Composing;
using Umbraco.Core.Logging;
using Umbraco.Core.Services;
using Umbraco.Web.Scheduling;

namespace Umbraco.Web.UI
{

>     public class CleanUpYourRoomComposer : ComponentComposer<CleanUpYourRoomComponent>

Reviewer: ```suggestion
    // We start by setting up a composer and component so our task runner gets registered on application startup
    public class CleanUpYourRoomComposer : ComponentComposer<CleanUpYourRoomComponent>
```

    {
    }

    public class CleanUpYourRoomComponent : IComponent
    {
        private IProfilingLogger _logger;
        private IRuntimeState _runtime;
        private IContentService _contentService;
        private BackgroundTaskRunner<IBackgroundTask> _cleanUpYourRoomRunner;

        public CleanUpYourRoomComponent(IProfilingLogger logger, IRuntimeState runtime, IContentService contentService)
        {
            _logger = logger;
            _runtime = runtime;
            _contentService = contentService;
            _cleanUpYourRoomRunner = new BackgroundTaskRunner<IBackgroundTask>("CleanYourRoom", _logger);
        }

        public void Initialize()
        {
            int delayBeforeWeStart = 60000; // 60000ms = 1min
            int howOftenWeRepeat = 300000; //300000ms = 5mins

            var task = new CleanRoom(_cleanUpYourRoomRunner, delayBeforeWeStart, howOftenWeRepeat, _runtime, _logger, _contentService);

            //As soon as we add our task to the runner it will start to run (after its delay period)
            _cleanUpYourRoomRunner.TryAdd(task);
        }

        public void Terminate()
        {
        }
    }

>     public class CleanRoom : RecurringTaskBase

Reviewer: ```suggestion
    // Now we get to define the recurring task
    public class CleanRoom : RecurringTaskBase
```

---
versionFrom: 8.0.0
---

# Scheduling with BackgroundTaskRunner

> In version Umbraco 8+ it is possible to run re-occuring code using the `BackgroundTaskRunner`.

Reviewer 2: ```suggestion
In Umbraco 8+ it is possible to run reccuring code using the `BackgroundTaskRunner`.
```

> Below is a complete example showing how to register the Task with the runner with a component that will regularly empty out the recycle bin every five minutes.

Reviewer 2: ```suggestion
Below is a complete example showing how to register a Task with the runner with a component that will regularly empty out the recycle bin every five minutes.
```
Will it empty out the recycle bin every five minutes, or regularly? 🤔

Author: Yes every five minutes, I know a bit over the top for an example but gives people some inspiration/ideas on what they could do hopefully @sofietoft
